### PR TITLE
Rework cmake and improvement protobuf

### PR DESCRIPTION
**Pull request by dudantas**
Cmake has been completely reworked, to work initially only with windows and linux, so that it replaces the visual studio solution and we can focus only on cmake, later I will create the pull request that depends on it, which is to add the github builds actions.

I improved the way protobuf was implemented, merging shared and appearances into one and making a proper lib for protobuf, thus removing the warnings and errors related to protobuf in the compilation.

It is necessary to test the compilation on linux, as I do not have this environment to be able to test

### CODE REVIEW

**Comment by beats-dh** [comment]
I can try to do actions and checks.

**Comment by dudantas** [comment]
> I can try to do actions and checks.

Thanks

**Comment by oarsc** [comment]
I don't have enough knowledge to review those changes. I can just compile it and see if it works correctly using cmake.
What I saw is that you removed the auto compiling .proto files changes that I made for cmake in #218 😢

**Comment by dudantas** [comment]
> I don't have enough knowledge to review those changes. I can just compile it and see if it works correctly using cmake.
> What I saw is that you removed the auto compiling .proto files changes that I made for cmake in #218 😢

Did not remove. I just replaced it with one that in my opinion works better (and without generating too many warnings). I don't want to be offensive, but the form I added is already used in canary and has worked well, even a guy on otclient discord tested my pull and reported that it solved the problems he had in relation to generating the protobuf.

**Comment by nekiro** [comment]
You are removing the vs solution to use cmake on windows, did you see any project doing that? Yeah I didn't too, there's a reason visual studio is used on windows and that's because its convenient and it's working fine, I see nothing "outdated" about it. Not to mention this change being completely unnecessary nor justified. What are we getting for ditching visual studio? Getting rid of few warnings that can be easily silenced or literally removed by implementing what I said in: https://github.com/mehah/otclient/pull/238#issuecomment-1073274175

I'm against these changes, but I guess my voice doesn't matter here much, anyway if maintainers decide to merge that then have fun with countless of questions on how to compile that and how does it work, in addition updating all the wiki, cause this completely changes the way project is compiled and what people are used to.

tl;dr a lot of changes, but I see no benefits

**Comment by Mrpox** [comment]
I totally agree with Nekiro on this.

**Comment by dudantas** [comment]
> You are removing the vs solution to use cmake on windows, did you see any project doing that? Yeah I didn't too, there's a reason visual studio is used on windows and that's because its convenient and it's working fine, I see nothing "outdated" about it. Not to mention this change being completely unnecessary nor justified. What are we getting for ditching visual studio? Getting rid of few warnings that can be easily silenced or literally removed by implementing what I said in: [#238 (comment)](https://github.com/mehah/otclient/pull/238#issuecomment-1073274175)
> 
> I'm against these changes, but I guess my voice doesn't matter here much, anyway if maintainers decide to merge that then have fun with countless of questions on how to compile that and how does it work, in addition updating all the wiki, cause this completely changes the way project is compiled and what people are used to.
> 
> tl;dr a lot of changes, but I see no benefits

Who said we're going to abandon visual studio? Did I say that? I don't intend to answer you, because of course you don't even know what you're talking about, you just want to rant at me, as usual!

image
image

The biggest difficulty you encountered was changing a single step on the wiki? If so, better look for other arguments! At least take the time to test before criticizing and saying what you don't know!

Not only will we have just one tool to compile all operating systems, windows, linux, and mac (and yes, it will continue to be visual studio, when you want to criticize something, at least take the time to test). And the compilation mode will remain the same, the only difference is that instead of the guy going there and opening the ".sln" he will open the "cmakelists.txt". Stop crying!

**Comment by dudantas** [comment]
Yes, I know several projects that use cmake: all. Especially if the goal is to maintain compatibility for several different operating systems. Even its heart project TFS, and the otc itself nowadays, uses cmake (for linux and mac). The difference here is that I put windows to use as well.

And please stop victimizing yourself, that's not cool!

**Comment by beats-dh** [comment]
Apart from that with vcpkg.json it installs all libs automatically, just open the project, of course it depends on having vcpkg on disk C: it is integrated with VS2022. Not to mention that the compilation on Linux is a disaster, it works but it can improve a lot, I made it as simple as possible you only need to install 2 libs on the outside and vcpkg does everything...

**Comment by Costallat** [comment]
> You are removing the vs solution to use cmake on windows, did you see any project doing that? Yeah I didn't too, there's a reason visual studio is used on windows and that's because its convenient and it's working fine, I see nothing "outdated" about it. Not to mention this change being completely unnecessary nor justified. What are we getting for ditching visual studio? Getting rid of few warnings that can be easily silenced or literally removed by implementing what I said in: [#238 (comment)](https://github.com/mehah/otclient/pull/238#issuecomment-1073274175)
> 
> I'm against these changes, but I guess my voice doesn't matter here much, anyway if maintainers decide to merge that then have fun with countless of questions on how to compile that and how does it work, in addition updating all the wiki, cause this completely changes the way project is compiled and what people are used to.
> 
> tl;dr a lot of changes, but I see no benefits

https://devblogs.microsoft.com/cppblog/cmake-support-in-visual-studio/

[Visual Studio 2017](https://aka.ms/vs2017cpp) introduces built-in support for handling CMake projects. This makes it a lot simpler to develop C++ projects built with CMake without the need to generate VS projects and solutions from the command line. This post gives you an overview of the CMake support, how to easily get started and stay productive in Visual Studio.

**Comment by dudantas** [comment]
https://github.com/mehah/otclient/pull/256
